native_client/deepspeech.cc

```diff
@@ -428,8 +494,87 @@ ModelState::decode(vector<float>& logits)
   vector<Output> out = ctc_beam_search_decoder(
     inputs.data(), n_frames, num_classes, *alphabet, beam_width,
     cutoff_prob, cutoff_top_n, scorer);
+	
+  return out;
+}
 
-  return strdup(alphabet->LabelsToString(out[0].tokens).c_str());
+vector<META_DICT>
+ModelState::metadata_from_output(Output out,int logit_count)
+{
+  vector<META_DICT> word_list;
+
+  const size_t num_classes = alphabet->GetSize() + 1; // +1 for blank
+  const int n_frames = logit_count / (BATCH_SIZE * num_classes);
+
+  std::string word = "";
+  int word_start_timestep = 0;
+
+  // Loop through each character
+  for (int t=0; t<out.tokens.size(); t++) {
+    const char * character = alphabet->StringFromLabel(out.tokens[t]).c_str();
+
+	if (strcmp(character," ") != 0) word.append(character);
+				
+	// Figure out word boundaries by looking for spaces
```

Comment:
Can you explain more about it being Latin-specific? Is there a way to make it language-independent? I’m not a linguist so I would need advice/help with that aspect.

Because if it has no concept of what a word is and only outputs per-letter timings, it’s only marginally more useful than the timesteps which are already exposed.

Maybe the API is the wrong place for it? Maybe it should be a helper function elsewhere or maybe it should be entirely client-side?

Reply:
> Maybe the API is the wrong place for it? Maybe it should be a helper function elsewhere or maybe it should be entirely client-side?

IMHO you should expose from the API only what is really needed. to produce the feature. The rest should either be helper or left to users.

I'm not against the idea of exposing some helpers from the API, but maybe @reuben has another point of view ?

Reply:
> Can you explain more about it being Latin-specific?

I guess relying on things like `strcmp(character, " ")` is going to be problematic for some languages.

Reply:
> I'm not against the idea of exposing some helpers from the API, but maybe @reuben has another point of view ?

Okay, I'm moving in the sense of @reuben: let's just expose a very low-level raw timing information. This should simplify your code in the `deepspeech.cc` lib.

Reply:
> I guess relying on things like `strcmp(character, " ")` is going to be problematic for some languages.

I still don't know what the exact issue is here. Is it that some languages use a different character code for spaces? Would std::isspace work? The docs say it detects whitespace for the _current_ locale, which would work assuming the user's locale is the same as the desired language, but I don't know if there is actually a way to reliably detect whitespace for any possible language you throw at it.

Also, the timesteps are already exposed in the Output structure, which would meet the definition of low-level raw timing information, so do I actually need to do anything further in the API at all? Maybe all of this should be client-side?

Reply:
Mandarin Chinese, for example, doesn't use spaces at all.

Reply:
The `Output` structure is currently not exposed in the API, so there's no way for clients to use it. My proposal is to move the word splitting code into the client and basically expose the `Output` structure in the API.

Reply:
Note that it can't literally be the existing `Output` struct though, since that uses std::vector and the native client API is C.

Reply:
> Mandarin Chinese, for example, doesn't use spaces at all.

Is there a solution to this?

Reply:
People would look at individual token timing rather than downsampling it like this. Hence why I think the API should expose raw token timings and let the client convert that into word-level, sentence level, whatever is needed by the application logic.

Reply:
Oh right, I keep thinking of it as a whole, rather than a separate API and client. Thanks @reuben, I will focus on getting those API changes approved first and work on the client after that.

Reply:
Just to clarify, I can only return C structures from the DS_xxx functions but it's ok to return C++ vectors from ModelState or StreamingState?

Reply:
Correct, `deepspeech.h` should be C only, everything else can use C++.